### PR TITLE
Add a Pythia8 gun for displaced particles (flat in Pt and Dxy)

### PR DESCRIPTION
**Pull request by mbluj**
#### PR description:

As title says: the PR adds a Pythia8 gun for displaced particles which are generated flat in Pt and Dxy in ranges provided by a user. This new gun enables generation of displaced unstable particles which is not possible with already existing [FlatRandomPtAndDxyGunProducer](https://github.com/cms-sw/cmssw/blob/master/IOMC/ParticleGuns/src/FlatRandomPtAndDxyGunProducer.cc).

Technically, the PR migrates an algorithm present in the [FlatRandomPtAndDxyGunProducer](https://github.com/cms-sw/cmssw/blob/master/IOMC/ParticleGuns/src/FlatRandomPtAndDxyGunProducer.cc) to a new module based on a generic Py8 gun ([Py8GunBase](https://github.com/cms-sw/cmssw/blob/master/GeneratorInterface/Pythia8Interface/interface/Py8GunBase.h)).

We consider to backport the PR to 11_2_X for a MC production for Run-3 studies.

#### PR validation:

It was validated with a custom workflow that the newly added module behaves as expected.  
The change does not affect any existing workflows by construction (a new "source" added).

### CODE REVIEW

**Comment by SiewYan** [comment]
please test

File: GeneratorInterface/Pythia8Interface/plugins/Py8PtAndDxyGun.cc [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        eta = (fMaxEta - fMinEta) * randomEngine().flat() + fMinEta;
+        double the = 2. * atan(exp(-eta));
+
+        mass = (fMasterGen->particleData).m0(particleID);
+
+        double pp = pt / sin(the);  // sqrt( ee*ee - mass*mass );
+        ee = sqrt(pp * pp + mass * mass);
+
+        pz = pp * cos(the);
+
```
[the remark this answers is not in the document]
Comment (mbluj) [reply GeneratorInterface/Pythia8Interface/plugins/Py8PtAndDxyGun.cc:107]: I adapted this code from [Py8PtGun](https://github.com/cms-sw/cmssw/blob/master/GeneratorInterface/Pythia8Interface/plugins/Py8PtGun.cc) plugin. I made a math and it is just another way to compute Cartesian 4-momentum components from pt/eta/phi (as expected).
After said this, I can change this to be more coherent with `FlatRandomPtAndDxyGunProducer` if it is your request for a price to be less coherent with other Py8 guns. Please, let me know.
BTW, I will update the module a bit to use more coherent naming of data members of the class.

[the remark this answers is not in the document]
Comment (SiewYan) [reply GeneratorInterface/Pythia8Interface/plugins/Py8PtAndDxyGun.cc:107]: Thanks! I overlooked Py8PtGun, the implementation looks ok to me.

**Comment by SiewYan** [comment]
please test

**Comment by SiewYan** [comment]
Looks fine to me. Thanks.

**Comment by SiewYan** [comment]
+1

**Comment by qliphy** [comment]
@mbluj   Could you also add a run test file?  You may refer to https://github.com/cms-sw/cmssw/pull/30759 and https://github.com/cms-sw/cmssw/pull/32227

**Comment by mbluj** [comment]
> 
> 
> @mbluj Could you also add a run test file? You may refer to #30759 and #32227

Sorry for belated answer. Yes, I will provide a test file, hopefully today.

**Comment by mbluj** [comment]
@qliphy a test cfg committed - it setup in a way to generate displaced taus. Please check if it is what you expect.

**Comment by qliphy** [comment]
@cms-sw/generators-l2   A test cfg has been added which looks good.  Would you please check and sign again if possible?  thanks!

**Comment by agrohsje** [comment]
+generators

**Comment by qliphy** [comment]
+1

**Comment by qliphy** [comment]
merge
added a test cfg since last test

**Pull request merged**